File: typescript-playwright/src/clients/AuthClient.ts
import type { AuthApi } from "../api/AuthApi";
import type { AuthResponse, LoginPayload } from "../utils/types";
import { statusIn } from "../utils/statusIn";
import { HTTP } from "../utils/constants";
import { step } from "../utils/decorators";

export class AuthClient {
  private readonly api: AuthApi;
 
  constructor(api: AuthApi) {
    this.api = api;
  }

  @step()
  async login(payload: LoginPayload): Promise<AuthResponse> {
    const response = await this.api.login(payload);
    if (!statusIn(HTTP.OK, HTTP.CREATED)(response.status())) {
      throw new Error(
        `[AuthClient] login failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<AuthResponse>;
  }

  @step()
  async refresh(refreshToken: string): Promise<AuthResponse> {
    const response = await this.api.refresh(refreshToken);
    if (!statusIn(HTTP.OK, HTTP.CREATED)(response.status())) {
      throw new Error(
        `[AuthClient] refresh failed: ${response.status()} ${await response.text()}`,
      );
    }
    return response.json() as Promise<AuthResponse>;
  }
}

File: python-pytest/src/clients/auth_client.py
from __future__ import annotations

from typing import cast

from src.api.auth_api import AuthApi
from src.utils.constants import HTTP
from src.utils.decorators import step
from src.utils.types import AuthResponse, LoginPayload


class AuthClient:
    def __init__(self, api: AuthApi) -> None:
        self._api = api

    @step()
    def login(self, payload: LoginPayload) -> AuthResponse:
        response = self._api.login(payload)
        if response.status_code not in (HTTP.OK, HTTP.CREATED):
            raise RuntimeError(f"[AuthClient] login failed: {response.status_code} {response.text}")
        return cast(AuthResponse, response.json())

    @step()
    def refresh(self, refresh_token: str) -> AuthResponse:
        response = self._api.refresh(refresh_token)
        if response.status_code not in (HTTP.OK, HTTP.CREATED):
            raise RuntimeError(
                f"[AuthClient] refresh failed: {response.status_code} {response.text}"
            )
        return cast(AuthResponse, response.json())
